pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -813,11 +828,14 @@ public void startSendOperation() {
         }
     }
 
-    public void completedSendOperation() {
+    public void completedSendOperation(boolean isNonPersistentTopic) {
         if (--pendingSendRequest == ResumeReadsThreshold) {
             // Resume reading from socket
             ctx.channel().config().setAutoRead(true);
         }
+        if (isNonPersistentTopic) {
+            nonPersistentMessageSemaphore.release();
```

Comment:
As I was mentioning above, the `pendingSendRequest` is already a counter of pending requests from the producer and it's not synchronized or a semaphore.